src/API.jl

```diff
@@ -975,7 +975,10 @@ function precompile(ctx::Context; internal_call::Bool=false)
                 end
                 is_direct_dep =  pkg in direct_deps
                 try
-                    !any(values(was_recompiled)) && printpkgstyle(ctx, :Precompiling, "project...")
+                    if !any(values(was_recompiled))
+                        was_recompiled[pkg] = true # needed to prevent `@async` race, and multiple prints
+                        printpkgstyle(ctx, :Precompiling, "project...")
```

Comment:
The tests in Base use a lock for printing:

https://github.com/JuliaLang/julia/blob/f6d1032763cfa160c117f0a5de25ba3370a3dca4/test/runtests.jl#L110